lib/sitemap-stream.ts

```diff
@@ -112,28 +175,70 @@ export class SitemapStream extends Transform {
     callback: TransformCallback
   ): void {
     if (!this.hasHeadOutput) {
+      // Add the opening tag size and closing tag size (since we have to close)
       this.hasHeadOutput = true;
-      this.push(getURLSetNs(this.xmlNS, this.xslUrl));
+      const headOutput = getURLSetNs(this.xmlNS, this.xslUrl);
+      this._byteCount += headOutput.length + closetag.length;
+      this.push(headOutput);
     }
-    if (
-      !this.smiStream.write(
-        validateSMIOptions(
-          normalizeURL(item, this.hostname, this.lastmodDateOnly),
-          this.level,
-          this.errorHandler
+
+    // Reject if item limit would be exceeded
+    if (this.countLimit !== undefined && this._itemCount === this.countLimit) {
+      // Write the close tag as the stream will be ended by raising an error
+      this.push(closetag);
+      this._wroteCloseTag = true;
+
+      callback(
+        new CountLimitExceededError(
+          'Item count limit would be exceeded, not writing, stream will close'
         )
+      );
+      return;
+    }
+
+    const itemOutput = SitemapItemToXMLString(
+      validateSMIOptions(
+        normalizeURL(item, this.hostname, this.lastmodDateOnly),
+        this.level,
+        this.errorHandler
       )
-    ) {
-      this.smiStream.once('drain', callback);
-    } else {
-      process.nextTick(callback);
+    );
+
+    // Check if the size would be exceeded by the new item
+    // and throw if it would exceed (when size limit enabled)
+    if (this.byteLimit !== undefined) {
+      if (
+        this._byteCount + Buffer.byteLength(itemOutput) >
+        this.byteLimit - closetagLen
```

Comment:
Oh, I don't think you want `- closetagLen` here.  I added that same fix myself on accident.

The `closetagLen` is already reserved inside the `if (!this.hasHeadOutput) {` block at the top of this `_transform` function, so you just have to check if adding this item length to the `_byteCount` will exceed the byte limit.  Subtracting out `closetagLen` here double-counts the length of the close tag.

Incidentally, we need to fix line 181 to use `closetagLen` instead of `closetag.length`

I think we also need `Buffer.byteLength` on `headOutput`?